solr/contrib/gcs-repository/src/java/org/apache/solr/gcs/GCSBackupRepository.java

```diff
@@ -0,0 +1,466 @@
+                            result.add(suffixName.substring(0, suffixName.length() - 1));
+                        } else {
+                            result.add(suffixName);
+                        }
+                    }
+                });
+
+        return result.toArray(new String[0]);
+    }
+
+    @Override
+    public IndexInput openInput(URI dirPath, String fileName, IOContext ctx) throws IOException {
+        return openInput(dirPath, fileName, ctx, 2 * 1024 * 1024);
+    }
+
+    private IndexInput openInput(URI dirPath, String fileName, IOContext ctx, int bufferSize) {
+        String blobName = dirPath.toString();
+        if (!blobName.endsWith("/")) {
+            blobName += "/";
+        }
+        blobName += fileName;
+
+        final BlobId blobId = BlobId.of(bucketName, blobName);
+        final Blob blob = storage.get(blobId, Storage.BlobGetOption.fields(Storage.BlobField.SIZE));
+        final ReadChannel readChannel = blob.reader();
+        readChannel.setChunkSize(bufferSize);
+
+        return new BufferedIndexInput(blobName, bufferSize) {
+
+            @Override
+            public long length() {
+                return blob.getSize();
+            }
+
+            @Override
+            protected void readInternal(ByteBuffer b) throws IOException {
+                readChannel.read(b);
+            }
+
+            @Override
+            protected void seekInternal(long pos) throws IOException {
+                readChannel.seek(pos);
+            }
+
+            @Override
+            public void close() throws IOException {
+                readChannel.close();
+            }
+        };
+    }
+
+    @Override
+    public OutputStream createOutput(URI path) throws IOException {
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucketName, toBlobName(path)).build();
+        final Storage.BlobWriteOption[] writeOptions = new Storage.BlobWriteOption[0];
+        final WriteChannel writeChannel = storage.writer(blobInfo, writeOptions);
+
+        return Channels.newOutputStream(new WritableByteChannel() {
+            @Override
+            public int write(ByteBuffer src) throws IOException {
+                return writeChannel.write(src);
+            }
+
+            @Override
+            public boolean isOpen() {
+                return writeChannel.isOpen();
+            }
+
+            @Override
+            public void close() throws IOException {
+                writeChannel.close();
+            }
+        });
+    }
+
+    @Override
+    public void createDirectory(URI path) throws IOException {
+        String name = path.toString();
+        if (!name.endsWith("/"))
+            name += "/";
+        storage.create(BlobInfo.newBuilder(bucketName, name).build()) ;
+    }
+
+    @Override
+    public void deleteDirectory(URI path) throws IOException {
+        List<BlobId> blobIds = allBlobsAtDir(path);
+        if (!blobIds.isEmpty()) {
+            storage.delete(blobIds);
```

Comment:
We delete the blobs inside path. Do we need to delete the directory itself?

Reply:
No, that should be implicitly included in the list that comes out of `allBlobsAtDir`.

This is a bit of a quirk that springs from GCS actually being a "flat" store that we're pretending has hierarchical structure like a filesystem.  GCS doesn't have a "list-directory" API since it doesn't really support directories.  Instead `allBlobsAtDir` retrieves all blobs whose names are prefixed with the "directory" path (which includes the directory being deleted itself).